sdk/identity/azure-identity/azure/identity/_credentials/on_behalf_of.py

```diff
@@ -65,14 +69,30 @@ class OnBehalfOfCredential(MsalCredential, GetTokenMixin):
             :caption: Create an OnBehalfOfCredential.
     """
 
-    def __init__(self, tenant_id: str, client_id: str, **kwargs: Any) -> None:
-        self._assertion = kwargs.pop("user_assertion", None)
+    def __init__(
+        self,
+        tenant_id: str,
+        client_id: str,
+        *,
+        client_certificate: Optional[bytes] = None,
+        client_secret: Optional[str] = None,
+        client_assertion_func: Optional[Callable[[], str]] = None,
+        user_assertion: str,
+        **kwargs: Any
+    ) -> None:
+        self._assertion = user_assertion
         if not self._assertion:
-            raise TypeError('"user_assertion" is required.')
-        client_certificate = kwargs.pop("client_certificate", None)
-        client_secret = kwargs.pop("client_secret", None)
+            raise TypeError('"user_assertion" must not be empty.')
 
-        if client_certificate:
+        if client_assertion_func:
+            if client_certificate or client_secret:
+                raise ValueError(
+                    'Specifying both "client_assertion_func" and "client_certificate" or "client_secret" is not valid.'
```

Comment:
This wording is a bit confusing. Should we reword to something like this?
```suggestion
                    'It is invalid to specify more than one of the following: "client_assertion_func", "client_certificate" or "client_secret".'
```

Reply:
if client_certificate or client_secret exists, should we ignore client_assertion_func to make it backward compat?

Reply:
I would consider that behavior unexpected. If the new argument is supplied, it might be surprising for it to be silently ignored.